frontend/src/views/Project/KmsPage/components/CmekTable.tsx

```diff
@@ -0,0 +1,409 @@
+import { useEffect, useState } from "react";
+import Link from "next/link";
+import {
+  faArrowDown,
+  faArrowUp,
+  faArrowUpRightFromSquare,
+  faCancel,
+  faCheckCircle,
+  faEdit,
+  faEllipsis,
+  faInfoCircle,
+  faKey,
+  faLock,
+  faLockOpen,
+  faMagnifyingGlass,
+  faPlus,
+  faTrash
+} from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { motion } from "framer-motion";
+
+import { createNotification } from "@app/components/notifications";
+import { ProjectPermissionCan } from "@app/components/permissions";
+import {
+  Badge,
+  Button,
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuItem,
+  DropdownMenuTrigger,
+  EmptyState,
+  IconButton,
+  Input,
+  Pagination,
+  Spinner,
+  Table,
+  TableContainer,
+  TableSkeleton,
+  TBody,
+  Td,
+  Th,
+  THead,
+  Tooltip,
+  Tr
+} from "@app/components/v2";
+import {
+  ProjectPermissionActions,
+  ProjectPermissionCryptoActions,
+  ProjectPermissionSub,
+  useProjectPermission,
+  useWorkspace
+} from "@app/context";
+import { useDebounce, usePopUp } from "@app/hooks";
+import { EncryptionAlgorithm, useGetCmeksByProjectId, useUpdateCmek } from "@app/hooks/api/cmeks";
+import { CmekOrderBy, TCmek } from "@app/hooks/api/cmeks/types";
+import { OrderByDirection } from "@app/hooks/api/generic/types";
+
+import { CmekDecryptModal } from "./CmekDecryptModal";
+import { CmekEncryptModal } from "./CmekEncryptModal";
+import { CmekModal } from "./CmekModal";
+import { DeleteCmekModal } from "./DeleteCmekModal";
+
+const INIT_PER_PAGE = 20;
+
+export const CmekTable = () => {
+  const { currentWorkspace } = useWorkspace();
+  const { permission } = useProjectPermission();
+
+  const [page, setPage] = useState(1);
+  const [perPage, setPerPage] = useState(INIT_PER_PAGE);
+  const [orderDirection, setOrderDirection] = useState(OrderByDirection.ASC);
+  const [search, setSearch] = useState("");
+  const debouncedSearch = useDebounce(search);
+
+  const projectId = currentWorkspace?.id ?? "";
+
+  const offset = (page - 1) * perPage;
+
+  const { data, isLoading, isFetching } = useGetCmeksByProjectId({
+    projectId,
+    offset,
+    limit: perPage,
+    search: debouncedSearch,
+    orderBy: CmekOrderBy.Slug,
+    orderDirection
+  });
+
+  const { keys = [], totalCount = 0 } = data ?? { keys: [], totalCount: 0 };
+
+  const { popUp, handlePopUpOpen, handlePopUpToggle } = usePopUp([
+    "upsertKey",
+    "deleteKey",
+    "encryptData",
+    "decryptData"
+  ] as const);
+
+  useEffect(() => {
```

Comment:
I am seeing this same conditions applied multiple times for pagination. I was thinking why not create a custom hook for pagination that has all the effects and page, perPage etc. This way we avoid declaring this each time everywhere

Reply:
refactored for all recent pagination implementations